### ISSUE TITLE
Feature ideas/requests

### ISSUE THREAD

**Issue by thesurlydev** [AndrewWalsh/openapi-devtools#6]
I left ideas in HN comments but leaving them here for posterity:

1. Ability to filter response properties.
2. Ability to work with non-JSON (web scraping) by defining a mapping of CSS selectors to response properties.
3. Cross-reference host names of captured requests with publicly documented APIs.
4. If auth headers are found, prompt the user for credentials that can then be stored locally.
5. "Repeater" functionality similar to the feature found in Burp Suite.
6. Generate clients on the fly based on the generated OpenAPI spec.
7. Train a machine learning model to recognize and extract tabular and repeated data based on training data.
8. Optionally publish generated OpenAPI specs to a central site or open PR to a GH repo, "awesome-openapi-devtools"?
9. Look for embedded data similar to what https://github.com/BishopFox/jsluice extracts and offer it as endpoints.

For some of these features, it may make sense to incorporate a server-side component. For example, sending captured data to be trained (7) or cross-referencing with existing documented APIs based on the host names (3).

**Comment by AndrewWalsh** [comment]
Thanks for relaying these suggestions and for your own @digitalsanctum, appreciate it:

1 & 2: I do want to keep this tool focused on doing the job it does well. There are quite a lot of existing extensions and options for scraping and doing similar such things. I think that use case might be best covered by a fork.

3: It's a client based tool without networking capabilities, so there's not much I can do here. Another issue is that "cross-reference" gets complicated quickly when publicly documented APIs don't match their behaviour. This use case is covered by the `Data Validators` section on [openapi.tools](https://openapi.tools/#data-validators), I don't think this tool is well placed to achieve this.

4: Agreed, will begin work on this shortly

5: The tool is not a HTTP interceptor as such and it doesn't modify requests or make them. I think this functionality would be brilliant, but it's not for this tool. A client-based HTTP interceptor that can modify requests/responses using this approach would have tremendous value, but it would need a full blown UI to go with it and would be non-trivial. Another issue is that this feature requires additional permissions.

6: I took that approach in https://atyourservice.awalsh.io/. Off the bat, it's not possible as it would mean pulling in [QuickType](https://www.npmjs.com/package/quicktype) and come with a large performance cost. What you can do though is manually copy response types into the [QuickType editor](https://app.quicktype.io/) if you need a one-off conversion. There are also many codegen tools for specific languages.

7: Great suggestion, I have some thoughts on that end. I can train a lean model to do that, I just need sufficient training data. LLMs and such are out, this would need to be a typical MLP using Tensorflow JS. Will think about this.

8: Nice idea, but to do that I'd need to introduce new permissions into the `manifest`. I'm keen to keep this extension focused, and without permission requirements.

9: Interesting request, I hadn't really thought about that. I will keep my eye on this prospect as you're right that URLs can contain a lot of information.

**Comment by AndrewWalsh** [comment]
@digitalsanctum I thought for a bit about your idea #6 to generate code. The simplest and best solution is to ensure users know they can use https://app.quicktype.io/ in combination with the app to generate code. I've made this option clear to users, thanks again for the suggestion.

I've also implemented #4.

I don't have time to implement #7 so I'll leave this for now, but I may come back to it in the future.

Thanks again for collating this, the codegen option in particular is a great idea that has contributed to the project.
